src/language/language_type.hbs.rs

```diff
@@ -53,22 +53,44 @@ impl LanguageType {
         ]
     }
 
-    /// Get language from it's file extension.
+    /// Get language from a file path. May open and read the file.
     ///
     /// ```no_run
     /// # use tokei::*;
-    /// let rust = LanguageType::from_extension("./main.rs");
+    /// let rust = LanguageType::from_path("./main.rs");
     ///
     /// assert_eq!(rust, Some(LanguageType::Rust));
     /// ```
-    pub fn from_extension<P: AsRef<Path>>(entry: P) -> Option<Self> {
-        if let Some(extension) = fs::get_extension(entry) {
+    pub fn from_path<P: AsRef<Path>>(entry: P) -> Option<Self> {
+        let filename = fs::get_filename(&entry);
+
+        if let Some(filename) = filename {
+            match &*filename {
+                {{~#each languages}}
+                    {{~#if this.filenames}}
+                        {{~#each this.filenames}}
+                            "{{~this}}" {{~#unless @last}} | {{~/unless}}
+                        {{~/each}}
+                            => return Some({{~@key}}),
+                    {{~/if}}
+                {{~/each}}
+
+                _ => ()
+            }
+        } 
+
+        let extension = fs::get_extension(&entry)
```

Comment:
I think that extension matching should be first. Just so it's a clear set of precedences in determining the language.
```
(extension) > (filename) > (shebang)
```